### DISCUSSION TITLE
How to know whether `Gt::Object_2` is a site or a line? (2D Apollonius diagram)

### DISCUSSION THREAD

**Discussion by stla** [CGAL/cgal#7580] [Q&A] [answered]
Hello,

As [written in the doc](https://doc.cgal.org/latest/Apollonius_graph_2/classCGAL_1_1Apollonius__graph__2.html#a304f5fe797914ddda090095ece7fdd7d), the dual of a face of an Apollonius graph has type `Gt::Object_2` and can be assignable to `Site_2` or `Gt::Line_2`. Currently in my code I do `Site2 site = ag.dual(f);` because I know the dual of each face is a site for my example (because the weights are distinct).

But how to do in general? That is, I'm looking for something like

```cpp
GT::Object_2 d = ag.dual(f);
if(typeof(d) == Site2) { 
  ......
} else {
  ......
}
```

I believe I have to use [`Assign_2`](https://doc.cgal.org/latest/Apollonius_graph_2/classApolloniusGraphTraits__2.html#a705fc3d395a7b7d0211b1c3f1e198a09) but I don't understand how it works.

**Comment by stla** [comment]
My attempts which do not work:

```cpp
    Object2 obj = ag.dual(f);
    Site2 site;

    if(Traits::assign_2_object(site, obj)) {
      Point2 pt  = site.point();
      DualPoints(fid, 0) = pt.x();
      DualPoints(fid, 1) = pt.y();
    } else {
      // ...
    }
```

```cpp
    Site2 site;

    if(Traits::assign_2_object(site, ag.dual(f))) {
      Point2 pt  = site.point();
      DualPoints(fid, 0) = pt.x();
      DualPoints(fid, 1) = pt.y();
    } else {
      // ...
    }
```

**Comment by stla** [comment, marked as answer]
This compiles so that probably works:

```cpp
    Traits traits = ag.geom_traits();
    Traits::Construct_object_2 cst = traits.construct_object_2_object();
    Object2 obj = cst(ag.dual(f));
    Traits::Assign_2 assgn = traits.assign_2_object();
    Site2 site;

    if(assgn(site, obj)) {
      Point2 pt  = site.point();
      DualPoints(fid, 0) = pt.x();
      DualPoints(fid, 1) = pt.y();
    } else {
      // ...
    }
```

**Reply by lrineau** [reply]
> This compiles so that probably works:
> 
> ```c++
>     Traits traits = ag.geom_traits();
>     Traits::Construct_object_2 cst = traits.construct_object_2_object();
>     Object2 obj = cst(ag.dual(f));
> ```

`ag.dual(f)` is already of type `Traits::Object_2`. The last line created a new `Object_2` from an existing object of the same type: that creates a copy. Harmless, but useless. You can write instead:

```c++
Traits::Object_2 obj = ag.dual(f);
```

And then:
> ```c++
>     Traits::Assign_2 assgn = traits.assign_2_object();
>     Site2 site;
> 
>     if(assgn(site, obj)) {
> ```

that is the correct API.

**Reply by lrineau** [reply]
Actually, the source of incomprehension is because the doc and code do not agree.

Here are the real signatures for `CGAL::Apollonius_graph_2::dual`:
https://github.com/CGAL/cgal/blob/6098856af6ad64f8c851d5e98416ccf4ea7b02a6/Apollonius_graph_2/include/CGAL/Apollonius_graph_2.h#L525-L529

See that the version from `Finite_faces_iterator` does not return an `Object_2` but directly a `Site_2`.

**Reply by stla** [reply]
Are you sure? `Traits::Object_2 obj = ag.dual(f);` is what I originally did, and this didn't compile (no conversion between `Site_2` and `Object_2`. And [the code](https://github.com/CGAL/cgal/blob/master/Apollonius_graph_2/include/CGAL/Apollonius_graph_2.h#L529) confirms what I say. Should we convert the `Finite_faces_iterator` `f` to a `Face_handle`?

**Reply by stla** [reply]
Cross posted. Yes I understand now. I can open an issue for the problem in the doc.

**Comment by stla** [comment]
Something weird happens. I replaced `ag.finite_faces` with `ag.all_faces` and then `assgn(site, obj)` is never true, as well as `assgn(line, obj)`. Will try to do a small reproducible example.

**Reply by stla** [reply]
And `std::cout << ag.dual(f)` always prints `1`.

**Comment by stla** [comment]
I checked <https://github.com/CGAL/cgal/blob/master/Apollonius_graph_2/include/CGAL/Apollonius_graph_2.h>. There's no method `dual` defined for `All_faces_iterator`. The doc claims there is one.
